subprojects/core/src/main/java/org/gradle/execution/DefaultWorkValidationWarningRecorder.java

```diff
@@ -36,10 +37,10 @@ public class DefaultWorkValidationWarningRecorder implements ValidateStep.Valida
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultWorkValidationWarningRecorder.class);
 
     // TODO If we can ensure that the recorder is only called once per work execution, then we can demote this to a simple counter
-    private final Set<UnitOfWork.Identity> workWithWarnings = ConcurrentHashMap.newKeySet();
+    private final Set<Identity> workWithWarnings = ConcurrentHashMap.newKeySet();
 
     @Override
-    public void recordValidationWarnings(UnitOfWork.Identity identity, UnitOfWork work, Collection<? extends InternalProblem> warnings) {
+    public void recordValidationWarnings(Identity identity, UnitOfWork work, Collection<? extends InternalProblem> warnings) {
         workWithWarnings.add(identity);
```

Comment:
🤔 Looking at this code makes me wonder whether all relevant `Identity` implementations are proper values wrt `equals` and `hashCode`. I know some aren't. Why not record the unique id instead?